### PR TITLE
Vested wallets deployment script

### PR DESCRIPTION
**Pull request by gvladika**
Add scripts for deploying and verifying vested wallets. Deployer script will create wallet factory contract, and then call `createWallets` function in a loop, 5 wallets at a time. There is no any token transfers involved. 

Prerequisites:
- .env containing: `ARB_KEY`, `ARB_URL`, `VESTED_RECIPIENTS_FILE_LOCATION` and `DEPLOYED_WALLETS_FILE_LOCATION`
- input list of recipients/beneficiaries should look like this (supports multiple wallets per beneficiary)
```
{
  "0xbf7258ead721d9ecc04a8476cf4f863f1b754497": ["1", "340"],
  "0xb98637f3750707fe6d1f0b35dbaf5fc8de65c63d": ["2"],
  "0x28e7A8CD861E9fd6D253bffE80B0704752Fd6A0D": ["650", "1100", "2500"]
}
```
- format of output with deployed addresses will look like this:
```
{
  "vestingWalletFactory": "0x24067223381F042fF36fb87818196dB4D2C56E9B",
  "beneficiaries": [
    {
      "beneficiary": "0xbf7258Ead721d9eCC04a8476Cf4F863F1b754497",
      "walletAddresses": [
        "0x10956D45F1d221D3b98898d0e7af28C20541057F",
        "0xA10AF745eC1245fC61ed63f891ce6ee5D8E57Ba3"
      ]
    },
    {
      "beneficiary": "0xb98637f3750707FE6d1F0b35dbAF5fC8De65c63d",
      "walletAddresses": [
        "0xE54D680B7DCA2eb6761d0797a5247bD03FB7DF16"
      ]
    },
    {
      "beneficiary": "0x28e7A8CD861E9fd6D253bffE80B0704752Fd6A0D",
      "walletAddresses": [
        "0x1e4e5273cb79a55296CDe3351925ad6FbdFFDfe1",
        "0x7afD37b828dA31ea4313fAD334A4B7A5A7832489",
        "0xAd47B4D46d927C6A1150107A26FC4286b08eB5F2"
      ]
    }
  ]
}
``` 

Deployment steps:
- prepare `.env`, ie. `cp files/goerli/.env-sample .env`
- prepare `vestingWalletRecipients.json ` in a location which is pointed to by `VESTED_RECIPIENTS_FILE_LOCATION `
- run deployed: `yarn deploy:vested-wallets`
- run verifier: `yarn verify:vested-wallets`

### CODE REVIEW

File: scripts/vestedWalletsDeployer.ts [hunk 1]
```diff
@@ -41,16 +76,60 @@ export const deployVestedWallets = async (
...
+    deployedWalletsLocation === undefined
+  ) {
+    throw new Error(
+      "Following env vars have to be defined: ARB_KEY, ARB_URL, VESTED_RECIPIENTS_FILE_LOCATION and DEPLOYED_WALLETS_FILE_LOCATION"
+    );
+  }
+
+  const deployedWalletsFileLocation = path.join(__dirname, "..", deployedWalletsLocation);
+  const vestedRecipientsFileLocation = path.join(__dirname, "..", vestedRecipientsLocation);
+  const vestedRecipients = loadRecipients(vestedRecipientsFileLocation);
```
Comment (DZGoldman) [inline scripts/vestedWalletsDeployer.ts:110]: how does this handle many wallets w/ the same beneficiary?

Comment (DZGoldman) [reply scripts/vestedWalletsDeployer.ts:110]: do we just run this more than once w/ two different recipient jsons?

File: scripts/vestedWalletsDeployer.ts [hunk 2]
```diff
@@ -41,16 +76,60 @@ export const deployVestedWallets = async (
...
+  const arbDeployer = new Wallet(arbKey, new JsonRpcProvider(arbRpc));
+
+  await deployVestedWallets(
+    deployedWalletsFileLocation,
+    arbDeployer,
+    vestedRecipients,
+    // start vesting in 1 years time
```
Comment (DZGoldman) [inline scripts/vestedWalletsDeployer.ts:117]:
```suggestion
    // start vesting 1 year after token generation 
```

**Review by DZGoldman — APPROVED** [review]
lgtm, just one clarifying question

File: scripts/vestedWalletsDeployer.ts (hunk 1 above)
Comment (gvladika) [reply scripts/vestedWalletsDeployer.ts:110]: hm yeah that would do it, although it would deploy 2 factories. Probably better if I modify data structure so each beneficiary has array of wallets instead of single one

Comment (gvladika) [reply scripts/vestedWalletsDeployer.ts]: done, updated PR description with new format for input and output

File: scripts/vestedWalletsDeployer.ts (hunk 2 above)
Comment (gvladika) [reply scripts/vestedWalletsDeployer.ts]: done

File: scripts/vestedWalletsDeploymentVerifier.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    assertNumbersEquals(
+      BigNumber.from(logs.length),
+      BigNumber.from(vestedRecipients[vr].length),
+      "Too many logs"
+    );
+
+    const log = logs[0];
+
+    const vestingWallet = ArbitrumVestingWallet__factory.connect(
```
Comment (yahgwai) [inline scripts/vestedWalletsDeploymentVerifier.ts:83]: Could be a separate PR, but would we could add a mode to this script that also checks that the correct token balance has actually been transferred.

**Review by yahgwai — APPROVED** [review]

File: scripts/vestedWalletsDeployer.ts [hunk 4]
```diff
@@ -1,31 +1,70 @@
...
   deployer: Signer,
-  tokenHolder: Signer,
-  tokenAddress: string,
-  recipients: Recipients,
+  recipients: VestedWallets,
   startTimeSeconds: number,
   durationSeconds: number
 ) => {
-  const token = L2ArbitrumToken__factory.connect(tokenAddress, tokenHolder);
+  let data = loadDeployedWallets(deployedWalletsFileLocation);
```
[the remark this answers is not in the document]
Comment (gvladika) [reply scripts/vestedWalletsDeployer.ts]: done

File: scripts/vestedWalletsDeployer.ts [hunk 5]
```diff
@@ -1,31 +1,70 @@
...
+    );
+  }
 
-  const recipientAddresses = Object.keys(recipients);
+  // make a list of beneficiary addresses where address can occur multiple times
+  const recipientAddresses = Object.keys(recipients)
+    .map((key) => Array(recipients[key].length).fill(key))
+    .flat();
+  // sum up number of already deployed wallets
+  const startIndex = data.beneficiaries.reduce((sum, obj) => sum + obj.walletAddresses.length, 0);
```
[the remark this answers is not in the document]
Comment (gvladika) [reply scripts/vestedWalletsDeployer.ts:60]: Added a section to readme with instructions and assumptions.

File: scripts/vestedWalletsDeploymentVerifier.ts (hunk 3 above)
Comment (gvladika) [reply scripts/vestedWalletsDeploymentVerifier.ts:83]: Right let's do it in separate PR, this one just focused on deploying and verifying the deployment.
